### PR TITLE
HPCC-17017 Add port info to socket tracename

### PR DESCRIPTION
**Pull request by mckellyln**
Signed-off-by: Mark Kelly <mark.kelly@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Full QA suite passed.

### CODE REVIEW

File: system/jlib/jsocket.cpp [hunk 1]
```diff
@@ -1531,7 +1546,17 @@ void CSocket::logPollError(unsigned revents, const char *rwstr)
...
+        try
+        {
+            lport = name(lname, sizeof(lname));
+        }
+        catch (IException *e)
+        {
+            strcpy(lname, "(NULL)");
+            lport = 1;
+            e->Release();
+        }
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Not related to original enhancement but discovered this and took the opportunity to add try/catch around name().

**Comment by richardkchapman** [comment]
@jakesmith Please review. Any performance concerns?

File: system/jlib/jsocket.cpp [hunk 2]
```diff
@@ -1469,10 +1469,25 @@ void CSocket::setTraceName(const char * prefix, const char * name)
...
+    }
+    peer.append(" -> ");
     peer.append(name?name:"(NULL)");
+    peer.append(":").append(hostport);
+    if (sock != INVALID_SOCKET)
+        peer.append(" (").append(sock).append(")");
 
     free(tracename);
-    tracename = strdup(peer);
+    tracename = strdup(peer.str());
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: Not new, but this would be better as:
```
tracename = peer.detach();
```

File: system/jlib/jsocket.cpp [hunk 3]
```diff
@@ -1469,10 +1469,25 @@ void CSocket::setTraceName(const char * prefix, const char * name)
...
+    }
+    catch (IException *e)
+    {
+        e->Release();
+    }
+    peer.append(" -> ");
     peer.append(name?name:"(NULL)");
+    peer.append(":").append(hostport);
+    if (sock != INVALID_SOCKET)
+        peer.append(" (").append(sock).append(")");
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: Doesn't feel like this should be done here (in setTraceName), but rather by the caller (Can common up elsewhere if same pattern).
e.g., CSocket::CSocket(const SocketEndpoint &ep, ... ,with name where trace name is "T>"
will have target port already and can use it.

File: system/jlib/jsocket.cpp [hunk 4]
```diff
@@ -1531,7 +1546,17 @@ void CSocket::logPollError(unsigned revents, const char *rwstr)
...
-        int lport = name(lname, sizeof(lname));
+        int lport;
+        try
+        {
+            lport = name(lname, sizeof(lname));
+        }
+        catch (IException *e)
+        {
+            strcpy(lname, "(NULL)");
+            lport = 1;
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: why 1?

File: system/jlib/jsocket.cpp (hunk 1 above)
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: what about other places that call name() , e.g. logConnectionInfo

Should name() always catch and behave like this?

File: system/jlib/jsocket.cpp [hunk 5]
```diff
@@ -2662,7 +2687,7 @@ CSocket::CSocket(const SocketEndpoint &ep,SOCKETMODE smode,const char *name)
         SocketEndpoint self;
         self.setLocalHost(0);
         self.getUrlStr(hostname);
-        setTraceName("S>", hostname);
+        setTraceName("S>", hostname.str());
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: fine, but not strictly necessary, const char * cast operator will call str() automatically.

**Review by jakesmith — CHANGES_REQUESTED** [review]

File: system/jlib/jsocket.cpp (hunk 4 above)
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: sorry, should be -1

File: system/jlib/jsocket.cpp (hunk 1 above)
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: logConnectionInfo may be ok as its always after a successful connect and if there is an error then throwing may make sense.  I will wrap it in try / catch and ignore error if you think that is better.
But in general case - should name() throw ?  I am not sure.

File: system/jlib/jsocket.cpp [hunk 6]
```diff
@@ -1469,10 +1469,25 @@ void CSocket::setTraceName(const char * prefix, const char * name)
...
+    }
+    peer.append(" -> ");
     peer.append(name?name:"(NULL)");
+    peer.append(":").append(hostport);
+    if (sock != INVALID_SOCKET)
+        peer.append(" (").append(sock).append(")");
 
     free(tracename);
-    tracename = strdup(peer);
+    tracename = peer.detach();
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: used detach()

File: system/jlib/jsocket.cpp [hunk 7]
```diff
@@ -2599,7 +2624,17 @@ void CSocket::set_ttl(unsigned _ttl)
...
+    try
+    {
+        lport = name(lname, sizeof(lname));
+    }
+    catch (IException *e)
+    {
+        strcpy(lname, "(NULL)");
+        lport = -1;
+        e->Release();
+    }
```
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: Added try / catch here also

File: system/jlib/jsocket.cpp (hunk 3 above)
Comment (mckellyln) [inline system/jlib/jsocket.cpp]: lport is local port, hostport is remote port, good to see both and sock fd in trace.

**Comment by mckellyln** [comment]
@jakesmith code updated as per review comments and discussion.

File: system/jlib/jsocket.cpp [hunk 8]
```diff
@@ -893,6 +894,8 @@ int CSocket::post_connect ()
         nagling = true;
         set_nagle(false);
         state = ss_open;
+        char lname[256];
+        localPort = name(lname, sizeof(lname), false);
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: a slight waste getting lname when not needed.
CSocket::name() should fill in a StringBuffer really too, which it does internally, only to copy to the fixed sized passed in buffer.
You could change it to 2 methods:
```
void CSocket::getEndpoint(SocketEndpoint &ep); 
unsigned CSocket::getName(StringBuffer &name);
```

and use getEndpoint() here.


Same is true of CSocket::peer_name()

Comment (jakesmith) [inline system/jlib/jsocket.cpp]: The oktothrow=false, is it anymore expect to throw an exception here than anywhere else?
The previously (prior to this commit) it would throw because called too early.

File: system/security/securesocket/securesocket.cpp [hunk 9]
```diff
@@ -234,9 +234,9 @@ class CSecureSocket : implements ISecureSocket, public CInterface
     }
 
     // Get name of accepted socket and returns port
-    virtual int name(char *name,size32_t namemax)
+    virtual int name(char *name, size32_t namemax, bool oktothrow=true)
```
Comment (jakesmith) [inline system/security/securesocket/securesocket.cpp]: minor: could use override

File: system/jlib/jsocket.cpp [hunk 10]
```diff
@@ -2691,9 +2704,12 @@ CSocket::CSocket(T_SOCKET new_sock,SOCKETMODE smode,bool _owned)
     accept_cancel_state = accept_not_cancelled;
     set_nagle(false);
     //set_linger(DEFAULT_LINGER_TIME); -- experiment with removing this as closesocket should still endevour to send outstanding data
-#ifdef _TRACE
     char peer[256];
-    peer_name(peer,sizeof(peer));
+    hostport = peer_name(peer,sizeof(peer));
+    hostname = strdup(peer);
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: can remove:
```
hostname = NULL;
hostport = 0;
```

early on in constructor

File: system/jlib/jsocket.cpp [hunk 11]
```diff
@@ -2691,9 +2704,12 @@ CSocket::CSocket(T_SOCKET new_sock,SOCKETMODE smode,bool _owned)
...
+    char lname[256];
+    localPort = name(lname, sizeof(lname), false);
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: It may not be an efficient concern, but as per prev. comment (in post_connect) it would look cleaner if method didn't get lname that it didn't need.

Unless it would be useful to store? I notice e.g. logConnectionInfo used to use, but you removed?

File: system/jlib/jsocket.cpp [hunk 12]
```diff
@@ -2598,11 +2614,7 @@ void CSocket::set_ttl(unsigned _ttl)
 
 void CSocket::logConnectionInfo(unsigned timeoutms, unsigned conn_mstime)
 {
-    char lname[256];
-    int lport = name(lname, sizeof(lname));
-    char rname[256];
-    int rport = peer_name(rname, sizeof(rname));
-    PROGLOG("SOCKTRACE: connect(%u) - time:%u ms fd:%d l:%s:%d r:%s:%d", timeoutms, conn_mstime, sock, lname, lport, rname, rport);
+    PROGLOG("SOCKTRACE: connect(%u) - time:%u ms fd:%d l:%d r:%s:%d", timeoutms, conn_mstime, sock, localPort, hostname, hostport);
```
Comment (jakesmith) [inline system/jlib/jsocket.cpp]: was logging lname not useful?

**Comment by jakesmith** [comment]
@mckellyln - a few comments.

Am not sure about the CSocket::name change (oktothrow).
see comment where oktothrow=false.
i.e. is it really more expected to throw an exception in that context than others?
If not, then why ignore in that 1 case not others.
Probably should throw there too (?) Socket is open by then now, so should succeed.
If should succeed (as with other contexts) then don't need oktothrow.

**Comment by mckellyln** [comment]
@jakesmith updated as per comments for re-review.

File: system/jlib/jsocket.hpp [hunk 13]
```diff
@@ -331,6 +331,9 @@ class jlib_decl ISocket : extends IInterface
     // Get peer ip of socket - in UDP returns return addr
     virtual IpAddress &getPeerAddress(IpAddress &addr)=0;
 
+    // Get local endpoint of socket
+    virtual SocketEndpoint &getEndpoint(SocketEndpoint &ep)=0;
```
Comment (jakesmith) [inline system/jlib/jsocket.hpp]: trivial: could be const method, but other existing methods (e.g. getPeerAddress) should be too.

File: system/security/securesocket/securesocket.cpp [hunk 14]
```diff
@@ -257,6 +257,12 @@ class CSecureSocket : implements ISecureSocket, public CInterface
         return m_socket->getPeerAddress(addr);
     }
 
+    // Get local endpoint of socket
+    virtual SocketEndpoint &getEndpoint(SocketEndpoint &ep)
```
Comment (jakesmith) [inline system/security/securesocket/securesocket.cpp]: could be const + override.

**Comment by jakesmith** [comment]
A couple of trivial comment, but looks good to merge now.
@richardkchapman

**Comment by mckellyln** [comment]
@jakesmith added const + override to new functions.  If you are ok with last commit, I will squash all commits.

**Comment by jakesmith** [comment]
@mckellyln - Yes looks good, go ahead and squash.

**Comment by mckellyln** [comment]
@jakesmith, @richardkchapman squashed and ready for merge.
